### ISSUE TITLE
False positive - Incompatible with protocol "Iterable"

### ISSUE THREAD

**Issue by nils-soderman** [microsoft/pyright#5433]
**Describe the bug**
If a class doesn't implement the `__iter__` dunder method it's considered incompatible with the type `Iterable` from the `typing` module. 

However a class that doesn't implement `__iter__` but implements `__getitem__` is still iterable as python will use `__getitem__` as a fallback if `__iter__` is not present.

**To Reproduce**
```python
from typing import Iterable

class MyClass:
    def __getitem__(self, index: int):
        return (0, 1, 2)[index]

def my_function(arg: Iterable[int]):
    for x in arg:
        print(x)

my_function(MyClass())
list(MyClass())
```

Code runs just fine, however 2 error messages will be generated when trying to use `MyClass()` in `my_function` & `list`

![image](https://github.com/microsoft/pyright/assets/20663273/46ea1c29-2532-4019-baa8-9b0b71d2361f)

**Expected behavior**
A class that implements `__getitem__` should be compatible with the `Iterable` type.

**VS Code extension or command-line**
VS Code 1.79.2
PyLance (ms-python.vscode-pylance) v2023.6.40
